hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -17,210 +17,198 @@
 
 from __future__ import absolute_import, division, print_function
 
-from hypothesis.internal.compat import hbytes, hrange
-from hypothesis.internal.conjecture.data import Status
+import attr
+
+from hypothesis.errors import Flaky, HypothesisException
+from hypothesis.internal.compat import hbytes, int_to_bytes
+from hypothesis.internal.conjecture.data import (
+    ConjectureData,
+    DataObserver,
+    Status,
+    StopTest,
+    bits_to_bytes,
+)
+
+
+class PreviouslyUnseenBehaviour(HypothesisException):
+    pass
+
+
+def inconsistent_generation():
+    raise Flaky(
+        "Inconsistent data generation! Data generation behaved differently "
+        "between different runs. Is your data generation depending on external "
+        "state?"
+    )
+
+
+EMPTY = frozenset()
+
+
+@attr.s(slots=True)
+class Branch(object):
+    bits = attr.ib()
+    children = attr.ib()
+
+
+@attr.s(slots=True, frozen=True)
+class Conclusion(object):
+    status = attr.ib()
+    interesting_origin = attr.ib()
+
+
+CONCLUSIONS = {}
+
+
+def conclusion(status, interesting_origin):
+    result = Conclusion(status, interesting_origin)
+    return CONCLUSIONS.setdefault(result, result)
+
+
+@attr.s(slots=True)
+class TreeNode(object):
+    """Node in a tree that corresponds to previous interactions with
+    a ``ConjectureData`` object according to some fixed test function.
+
+    This is functionally a variant patricia trie.
+    See https://en.wikipedia.org/wiki/Radix_tree for the general idea,
+    but what this means in particular here is that we have a very deep
+    but very lightly branching tree and rather than store this as a fully
+    recursive structure we flatten prefixes and long branches into
+    lists. This significantly compacts the storage requirements.
+    """
+
+    # Records the previously drawn bits and their
+    # corresponding values. Either len(bits) == len(values)
+    # or len(bits) == len(values) + 1 and the last bit call
+    # corresponds to the transition to a child node.
+    bits = attr.ib(default=attr.Factory(list))
```

Comment:
This could use a comment and/or a renaming to clarify that its values are bit-lengths (i.e. arguments to `draw_bits`), since it’s not obvious to an unfamiliar reader.

Reply:
Done.